### PR TITLE
refactor and test `ui_focus_system`

### PR DESCRIPTION
**Pull request by pubrrr**
# Objective

While digging through the bevy code, I found that `ui_focus_system` overly complex and quite hard to understand. -> make it easier to understand.

## Solution

- write unit tests
- refactor
- I tried to introduce no logical changes

### CODE REVIEW

**Comment by pubrrr** [comment]
Ok, apparently the ci fails:

     = rustc_version v0.2.3
      └── (build) stdweb v0.4.20
          └── gilrs-core v0.3.2
     = rustc_version v0.4.0
      └── (build) rstest v0.12.0
          └── (dev) bevy_ui v0.7.0-dev

Is there any chance to make this work?

**Comment by alice-i-cecile** [comment]
For the CI, the workaround is to whitelist the duplicate dependencies. That said, I'm reluctant to pull a unit testing framework into our dependency tree just for this.

What benefit does it offer here?

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/focus.rs:59]: I don't think this actually makes the code clearer, and it *definitely* shouldn't be `pub`.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/focus.rs:77]: I don't think it's useful to have a system whose contents are a single function call: that seems like needless layering.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/focus.rs:93]: I like this living in a separate system at least.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/focus.rs:101]: I think you should be able to use the ? operator here.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/focus.rs:131]: This feels like it should be a method on Node or something to that effect?

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/focus.rs:209]: I really dislike the use of the generic plus trait here.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/focus.rs:222]: I don't think that this adds enough value to pull in a testing library on a one-off basis. These tests should be feasibel to rewrite with a helper method instead.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/focus.rs:221]: These tests look good, and we definitely need tests for this piece of code.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Overall, I think that this region could use code quality improvements to break up the mega-function, and it definitely needs tests!

I pretty strongly dislike rstest here (I don't think it's clearer, and it pulls in a new dependency). The refactor for the code itself needs a bit more iteration, but it feels like breaking this out into smaller pieces is the right path.

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs:59]: I mainly introduced it so that we don't need this in 3 different places. But I'm also not really satisfied as it makes it hard to see what is actually in the query. Should I revert it?

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs:77]: This is to make it testable. Or do you have other suggestions how to mock `Windows`?

I.e. `ui_focus_system` is actually meant as a public export of `focus_ui::<Windows>`

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/focus.rs:59]: I think revert it, and then we can see if there are other strategies to avoid reuse.

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs:131]: I also thought that this function looks a bit odd as a standalone. But is `Node` the right place as `Node` is actually indifferent to positions?

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/focus.rs:77]: Ah. I think we should try to tackle that directly: I don't want to have to replicate this strategy everywhere.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/focus.rs:209]: Ah I see: it's to work around the difficulty mocking `Windows`?

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs:221]: Is there actually a common practice how to test systems? This TestApp approach is something I came up with in one of my own projects from the Cheatbook suggesting Direct World Access. But I didn't see many tests around in bevy itself.

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs:209]: See my other comment - how else would I make `Windows` do what I want in a unit test? Everything I tried ended up with a dependency on hardware.

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs:93]: Should I extract it? This system currently does two things:
- resetting Interactions to None
- checking the input to set some Interactions to Hovered / Clicked

The coupling between those two things is quite hard as they rely on each other. This becomes quite obvious from the `Local<State>`. Actually I wanted to get rid of it, but I did not have a good idea without changing the logic.

**Comment by alice-i-cecile** [comment]
So, WRT the testing strategy: #3835 is the blocker here. IMO we should hold off on this until that's merged, then use the headless windows there to enable us to write proper tests for this code.

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs:101]: The compiler says: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `FromResidual`)

Or do you mean I should return an Option here?

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs:59]: done

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs]: I kinda handcoded it myself now

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [inline crates/bevy_ui/src/focus.rs:217]: - [ ] I might need to be careful here concerning change detection. Double check.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/focus.rs:116]: Ah my bad; I'd misread what the function does.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/focus.rs:243]: You can see how I did it in one of my external crates here: https://github.com/Leafwing-Studios/leafwing_input_manager/blob/main/tests/integration.rs

File: crates/bevy_ui/Cargo.toml
Comment (alice-i-cecile) [inline crates/bevy_ui/Cargo.toml:35]: We should revert this change to avoid noise; it looks like you're missing a trailing newline.

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/focus.rs]: Excellent, thanks!

File: crates/bevy_ui/src/focus.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/focus.rs:146]: Yeah 🤔 TBH, this feels like functionality that should be pub that we reuse internally, but I can't immediately figure out the best design for that.

**Comment by nicopap** [comment]
You might be interested in https://github.com/bevyengine/rfcs/pull/41 and https://github.com/nicopap/ui-navigation. I'm planning to start the bevy PR next week (I still need to update the RFC with the second round of changes). I'm not sold on `rtest` either, to be perfectly honest, the test code was completely obtuse to me :grinning:. Looks great now though! 

The `ui-navigation` crate is sorely lacking when it comes to tests. And unlike bevy's focus system, it's fairly well documented.

File: crates/bevy_ui/src/focus.rs
Comment (pubrrr) [reply crates/bevy_ui/src/focus.rs:217]: I guess after adding a test that checks change detection, unfortunately most of my changes won't work.

At least there are a couple more tests. The `Option<Interaction>` change can be kept, and maybe a refactoring of cutting it into smaller functions. But the approach "set everything to `None`, then check the state" won't work.

**Comment by pubrrr** [comment]
I guess after adding a test that checks change detection, unfortunately most of my changes won't work.

At least there are a couple more tests. The "`Option<Interaction>` -> `Interaction`" change can be kept, and maybe a refactoring of cutting it into smaller functions. But the approach "set everything to None, then check the state" won't work. Or does anyone have an idea how to make it work?

Anyway, I'll leave this PR around. Ping me once headless windows are there. I believe the tests can easily be adapted to work without the extra trait then.

**Comment by NthTensor** [comment]
This looks like it has gone stale, and we are now planning to deprecate `Interaction` anyway. Closing as not planned.

**Pull request closed without merging**
